### PR TITLE
Support Aliases for nodes 

### LINKED ISSUE #718: Feature request: Add DNS A records of nodes in Route53 via dns-controller
### Use case

If we're running a UDP service (like StatsD), and we need to expose it publicly, we can use only `NodePorts` since `ELBs` don't work with UDP.
### Solution

One possible solution is keeping updated A records for all the nodes in Route53, and using a pre-defined `nodePort` value.

So, it would be awesome if `dns-controller` can create and keep the records up-to-date with the IPs of healthy nodes. Extra options like setting a TTL value would also be nice to have.

cc: @justinsb

### PR DESCRIPTION
**Pull request by ApsOps**
Fixes #718 

Based on PR #720 

cc: @justinsb

### CODE REVIEW

**Comment by chrislovecnm** [comment]
Please sign the Linuxfoundation CLA

**Comment by ApsOps** [comment]
@chrislovecnm I did sign it. Not sure why it still isn't showing up here.

Yeah, I think we should add tests, but I'm golang beginner so I'll probably need some help. I'll try it out anyway.

**Comment by chrislovecnm** [comment]
Did you sign the google cla as well?

File: dns-controller/pkg/dns/dnscontroller.go [hunk 1]
```diff
@@ -22,13 +22,14 @@ import (
```
Comment (chrislovecnm) [inline dns-controller/pkg/dns/dnscontroller.go]: Some unit tests :)

**Comment by ApsOps** [comment]
@chrislovecnm I had signed a Google cla for a PR to main k8s repo. It's the same cla I believe?

**Comment by chrislovecnm** [comment]
yep ... let me remove the label, and the next time you push an update to the PR, the bot should check again.

**Comment by ApsOps** [comment]
@chrislovecnm Added a few unit tests. Please let me know if something is wrong :)

File: dns-controller/pkg/dns/zonespec_test.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+// This is not correct
+
+// func TestMatchesExplicitly(t *testing.T) {
+// 	cases := []struct {
+// 		r        ZoneRules
+// 		zone     dnsprovider.Zone
+// 		expected bool
+// 	}{
+// 		{
+// 			ZoneRules{
```
Comment (ApsOps) [inline dns-controller/pkg/dns/zonespec_test.go]: Not sure how to test functions with receiver arguments (methods). Can someone please point to an example?

File: dns-controller/pkg/dns/zonespec_test.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+					&ZoneSpec{Name: "example.com.", ID: "1234"},
+				},
+				Wildcard: false,
+			},
+		},
+	}
+
+	for _, c := range cases {
+		if actual, _ := ParseZoneRules(c.zones); !reflect.DeepEqual(*actual, c.expected) {
+			t.Errorf("ParseZoneRules(%#v) expected %#v, but got %#v", c.zones, c.expected, *actual)
```
Comment (ApsOps) [inline dns-controller/pkg/dns/zonespec_test.go]: Failing tests don't print the exact value here (prints just the pointer address) since it's a nested pointer-struct.
Is there a way to print the exact value here?

File: dns-controller/pkg/dns/zonespec_test.go (hunk 2 above)
Comment (chrislovecnm) [inline dns-controller/pkg/dns/zonespec_test.go]: @kris-nova / @justinsb / @zmerlynn ideas?

File: dns-controller/pkg/dns/zonespec_test.go (hunk 3 above)
Comment (chrislovecnm) [inline dns-controller/pkg/dns/zonespec_test.go]: `%+v`?

**Comment by chrislovecnm** [comment]
Looks great.  If you are unable to determine how to test it, comment in the code and file and issue.  We will want https://github.com/kubernetes/kops/blob/master/Makefile#L80 <- update to do the test :)

**Comment by chrislovecnm** [comment]
@k8s-ci-robot ok test this

**Comment by chrislovecnm** [comment]
@k8s-bot aws e2e test this

@zmerlynn this is another edge case.  When we work on protokube or dnscontroller we cannot really do a e2e test :(  With this PR do you want to move dnscontroller into gcr.io?

File: dns-controller/pkg/dns/zonespec_test.go (hunk 3 above)
Comment (chrislovecnm) [inline dns-controller/pkg/dns/zonespec_test.go]: good enough.  You would need to do a pretty print.

**Comment by chrislovecnm** [comment]
LGTM

**Comment by ApsOps** [comment]
@chrislovecnm Updated README.

**Comment by chrislovecnm** [comment]
Wahoo!!
